app/models/location_report.rb

```diff
@@ -0,0 +1,208 @@
+  validates :end_date, presence: { if: :report_type_selection?, date: { after_or_equal_to: :start_date } }
+
+  validates :barcodes_text, presence: { if: :report_type_barcodes? }
+  validate :barcodes_are_recognised, if: :report_type_barcodes?
+  validate :plates_are_found, if: :report_type_selection?
+
+  def report_type_selection?
+    report_type == 'selection'
+  end
+
+  def report_type_barcodes?
+    report_type == 'barcodes'
+  end
+
+  def barcodes_are_recognised
+    if barcodes_list.blank?
+      errors.add(:barcodes_text, 'Please enter some barcodes in the text area')
+      return
+    end
+
+    ean13_barcodes_list = []
+    invalid_barcodes_list = []
+    barcodes_list.each do |cur_bc|
+      if barcode_is_human_readable?(cur_bc)
+        # human readable format
+        ean13_barcodes_list.push(Barcode.human_to_machine_barcode(cur_bc))
+      elsif barcode_is_ean13?(cur_bc)
+        # already ean 13 format so unchanged
+        ean13_barcodes_list.push(cur_bc)
+      else
+        invalid_barcodes_list.push(cur_bc)
+      end
+    end
+
+    unless invalid_barcodes_list.size.zero?
+      errors.add(:barcodes_text, "Invalid barcodes found: #{invalid_barcodes_list.join(',')}")
+      return
+    end
+
+    if ean13_barcodes_list.size.zero?
+      errors.add(:barcodes_text, 'Please enter some valid barcodes (human or scannable) in the text area')
+      return
+    end
+
+    self.barcodes = ean13_barcodes_list.join(' ')
+  end
+
+  def plates_are_found
+    unless Plate.search_for_plates(
+      study_id:             study_id,
+      start_date:           start_date,
+      end_date:             end_date,
+      plate_purpose_ids:    plate_purpose_ids
+    ).any?
+      errors.add(:study_id, 'Those selection criteria return no plates')
+    end
+  end
+
+  def schedule_report
+    Delayed::Job.enqueue LocationReportJob.new(id)
+  end
+
+  def generate_report_rows
+    plates_list = []
+
+    if report_type == 'barcodes'
+      ean13_barcodes_list = []
+      self.barcodes_list = barcodes.strip.split(/[\s]+/)
+      if barcodes_list.empty?
+        yield(['Error: barcodes list was empty when generating the report'])
+        return
+      end
+      barcodes_list.each do |cur_bc|
+        if barcode_is_human_readable?(cur_bc)
+          # human readable format
+          ean13_barcodes_list.push(Barcode.human_to_machine_barcode(cur_bc))
+        elsif barcode_is_ean13?(cur_bc)
+          # ean 13 format so unchanged
+          ean13_barcodes_list.push(cur_bc)
+        end
+      end
+
+      # fetch the plates for the batch of barcodes
+      plates_list = Plate.with_machine_barcode(ean13_barcodes_list)
+    elsif report_type == 'selection'
+      plates_list = search_for_plates_by_selection
+    end
+
+    if plates_list.empty?
+      yield(['Error: plates list was empty when generating the report'])
+      return
+    end
+
+    yield headers
+
+    plates_list.each do |cur_plate|
```

Comment:
Pull into separate method, pass block through with &block (add as final argument to outer method)